src/preamble.js

```diff
@@ -2318,19 +2318,38 @@ function integrateWasmJS() {
     // later), so we save Module and check it later.
     var trueModule = Module;
 #endif
-    getBinaryPromise().then(function(binary) {
-      return WebAssembly.instantiate(binary, info)
-    }).then(function(output) {
+    function receiveInstantiatedSource(output) {
+      // 'output' is a WebAssemblyInstantiatedSource object which has both the module and instance.
       // receiveInstance() will swap in the exports (to Module.asm) so they can be called
 #if ASSERTIONS
       assert(Module === trueModule, 'the Module object should not be replaced during async compilation - perhaps the order of HTML elements is wrong?');
       trueModule = null;
 #endif
       receiveInstance(output['instance']);
-    }).catch(function(reason) {
-      Module['printErr']('failed to asynchronously prepare wasm: ' + reason);
-      abort(reason);
-    });
+    }
+    function instantiateArrayBuffer(receiver) {
+      getBinaryPromise().then(function(binary) {
+        return WebAssembly.instantiate(binary, info);
+      }).then(receiver).catch(function(reason) {
+        Module['printErr']('failed to asynchronously prepare wasm: ' + reason);
+        abort(reason);
+      });
+    }
+    // Prefer streaming instantiation if available.
+    if (!Module['wasmBinary'] && typeof WebAssembly.instantiateStreaming === 'function') {
+      WebAssembly.instantiateStreaming(fetch(wasmBinaryFile, { credentials: 'same-origin' }), info)
+          .then(receiveInstantiatedSource).catch(
+              function(reason) {
+                // We expect the most common failure cause to be a bad MIME type for the binary,
+                // in which case falling back to ArrayBuffer instantiation should work.
+                Module['printErr']('wasm streaming compile failed: ' + reason);
+                Module['printErr']('falling back to ArrayBuffer instantiation');
+                instantiateArrayBuffer(receiveInstantiatedSource);
+              }
+      );
+      return {};
+    }
+    instantiateArrayBuffer(receiveInstantiatedSource);
```

Comment:
Shouldn't this line be in `else` branch?

Reply:
The previous `if` ends with a `return`, so no need.

Reply:
I'd prefer `else` here, since there is one more `return {};` one line below, now it is duplicated in 2 places.

Reply:
Additional nit: if we invert the conditional to put the shorter branch first, that's probably more readable whether we early-return or if-else.
Not sure if the condition is more readable as `!Module['wasmBinary'] && typeof WebAssembly.instantiateStreaming === 'function'` or `Module['wasmBinary'] || typeof WebAssembly.instantiateStreaming !== 'function'`, probably the former? So, `var canStream = ...; if (!canStream) { ...`

Reply:
I'm a little torn on that because I usually prefer positive conditions ("we can stream") and to put the "expected" case first and the unusual/exceptional case second. I went ahead and put the else in since there's only 2 cases and to deduplicate the return. How does it look now?

Reply:
In any case based on the discussion in #5577 we may end up with a different condition anyway...

Reply:
I feel like that's an argument for early-return, more so than for shorter conditionals. I think in practice that does often mean short error cases go first and get early-returns, like in your linked example. But that's still "if (positive case) then error" and I want "if (positive case) then stream". I do think that if we had more than just 2 cases then the order would probably be determined by how they composed and/or how readable each condition was. With just 2 fairly-short cases it's less obvious. And also the stakes are much lower, so :woman_shrugging: